File: python/Q981_time_based_key_value_store.py
#
# @lc app=leetcode id=981 lang=python3
#
# [981] Time Based Key-Value Store
#


# @lc code=start
class TimeMap:
    def __init__(self) -> None:
        self.values: dict[str, list[tuple[str, int]]] = dict()

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.values:
            self.values[key] = []
        self.values[key].append((value, timestamp))

    def get(self, key: str, timestamp: int) -> str:
        values: list[tuple[str, int]] = self.values.get(key, [])

        left: int = 0
        right: int = len(values) - 1

        while left <= right:
            m: int = left + ((right - left) // 2)

            if values[m][1] <= timestamp:
                left = m + 1
            else:
                right = m - 1

        return values[right][0] if right >= 0 else ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
# @lc code=end

File: rust/src/leetcode/Q981_time_based_key_value_store.rs
/*
 * @lc app=leetcode id=981 lang=rust
 *
 * [981] Time Based Key-Value Store
 */

// @lc code=start
use std::collections::HashMap;

struct TimeMap {
    map: HashMap<String, Vec<(String, i32)>>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl TimeMap {
    fn new() -> Self {
        TimeMap {
            map: HashMap::new(),
        }
    }

    fn set(&mut self, key: String, value: String, timestamp: i32) {
        self.map
            .entry(key)
            .or_insert(vec![])
            .push((value, timestamp));
    }

    fn get(&self, key: String, timestamp: i32) -> String {
        static EMPTY: Vec<(String, i32)> = vec![];
        let saved = self.map.get(&key).unwrap_or(&EMPTY);

        let mut left: i32 = 0;
        let mut right: i32 = saved.len() as i32 - 1;

        while left <= right {
            let mid: i32 = left + (right - left) / 2;

            if saved[mid as usize].1 <= timestamp {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        if right >= 0 {
            saved[right as usize].0.clone()
        } else {
            String::new()
        }
    }
}

// /**
//  * Your TimeMap object will be instantiated and called as such:
//  * let obj = TimeMap::new();
//  * obj.set(key, value, timestamp);
//  * let ret_2: String = obj.get(key, timestamp);
//  */
// @lc code=end
